definitions/npm/firebase_v4.x.x/flow_v0.34.x-v0.45.x/firebase_v4.x.x.js

```diff
@@ -1,344 +1,402 @@
+  ): Promise<void>,
+  update(
+    values: { [path: string]: FirebaseValue },
+    onComplete?: OnCompleteCallback
+  ): Promise<void>
 }
 
+declare type FirebaseDatabaseCallback = (DataSnapshot, ?string) => void
+
 declare class Query {
-  ref: Reference;
-  endAt(value: number | string | boolean | null, key?: string): Query;
-  equalTo(value: number | string | boolean | null, key?: string): Query;
-  isEqual(other: Query): boolean;
-  limitToFirst(limit: number): Query;
-  limitToLast(limit: number): Query;
+  ref: Reference,
+  endAt(value: number | string | boolean | null, key?: string): Query,
+  equalTo(value: number | string | boolean | null, key?: string): Query,
+  isEqual(other: Query): boolean,
+  limitToFirst(limit: number): Query,
+  limitToLast(limit: number): Query,
   off(
-    eventType: QueryEventType,
-    callback: ?(DataSnapshot, ?string) => void,
-    context: Object,
-  ): void;
+    eventType?: QueryEventType,
+    callback?: FirebaseDatabaseCallback,
+    context?: Object
+  ): void,
   on(
     eventType: QueryEventType,
-    callback: (DataSnapshot, ?string) => void,
+    callback: FirebaseDatabaseCallback,
     cancelCallbackOrContext?: (error: Object) => void | Object,
-    context?: (DataSnapshot, ?string) => void,
-  ): (DataSnapshot, ?string) => void;
+    context?: FirebaseDatabaseCallback
+  ): FirebaseDatabaseCallback,
   once(
     eventType: QueryEventType,
-    successCallback?: (DataSnapshot, ?string) => void,
+    successCallback?: FirebaseDatabaseCallback,
     failureCallbackOrContext?: (error: Object) => void | Object,
-    context?: Object,
-  ): Promise<DataSnapshot>;
-  orderByChild(path: string): Query;
-  orderByKey(): Query;
-  orderByPriority(): Query;
-  orderByValue(): Query;
-  startAt(value: number | string | boolean | null, key?: string): Query;
-  toJSON(): Object;
-  toString(): string;
+    context?: Object
+  ): Promise<DataSnapshot>,
+  orderByChild(path: string): Query,
+  orderByKey(): Query,
+  orderByPriority(): Query,
+  orderByValue(): Query,
+  startAt(value: number | string | boolean | null, key?: string): Query,
+  toJSON(): Object,
+  toString(): string
 }
 
 declare class Reference extends Query {
-  key: ?string;
-  parent?: Reference;
-  root: Reference;
-  child(path: string): Reference;
-  onDisconnect(): OnDisconnect;
-  push(value?: FirebaseValue, onComplete?: OnCompleteCallback): ThenableReference;
-  remove(onComplete?: OnCompleteCallback): Promise<void>;
-  set(value: FirebaseValue, onComplete?: OnCompleteCallback): Promise<void>;
-  setPriority(priority: Priority, onComplete?: OnCompleteCallback): Promise<void>;
-  setWithPriority(newVal: FirebaseValue, newPriority: Priority, onComplete?: OnCompleteCallback): Promise<void>;
+  key: ?string,
+  parent?: Reference,
+  root: Reference,
+  child(path: string): Reference,
+  onDisconnect(): OnDisconnect,
+  push(
+    value?: FirebaseValue,
+    onComplete?: OnCompleteCallback
+  ): ThenableReference & Promise<void>,
```

Comment:
You mentioned that in your comments that it should be ThenableReference or Promise. Should this be |?